keras_nlp/applications/bert.py

```diff
@@ -198,3 +199,46 @@ def get_config(self):
             }
         )
         return config
+
+
+class BertClassifier(keras.Model):
```

Comment:
I think this too should be a functional model.

Reply:
Also, would be great to have a way to distinguish the "head adapters" from the "architectures" themselves in the API symbol name. The usage is fairly different.

We need this in a way that would scale across all the different types of models we have (so not specific to encoders). Any ideas?

Reply:
For a functional model, is there a non-awkward way to sync it with the fairly complex inputs of the encoder? Otherwise it could be hard to keep them in sync.

Reply:
I feel like the head args make it pretty clear that the design pattern is composition: a `BertClassifier` has a `Bert` encoder. I might need more guidance on what we're trying to clarify.

Reply:
For the second question. I'm not saying I like these names below, but conceptually it seems like it might be nice to have something like this:

```
keras_nlp.models.BertLarge()
keras_nlp.models.BertBase()
keras_nlp.models.adapters.BertSequenceClassifier()
keras_nlp.models.adapters.BertTokenClassifier()
keras_nlp.models.adapters.BertSequenceRegressor()
keras_nlp.models.adapters.BertSpanLabeler()

# or
keras_nlp.models.BertLarge()
keras_nlp.models.BertBase()
keras_nlp.models.BertSequenceClassificationHead()
keras_nlp.models.BertTokenClassifationHead()
keras_nlp.models.BertSequenceRegressionHead()
keras_nlp.models.BertSpanLabelingHead()

#or ...
```

Basically, a way to tell at a glance, as a user, whether the object they are seeing should be initialized directly, or should be passed a base architecture as the first argument. (Side note, might be nice to expose the base architecture on the adapted class in a consistent way. `model.base_model` or something like that).

Right now, we are shoving two distinct set of objects, will somewhat different usage patterns into the exact same name space. That may be the best solution, but wanted to at least think about an alternative naming scheme we could go with.

Reply:
I think the current naming is clear. `BertClassifier` is not a `head`, it is a full model with a built-in encoder for raw token inputs.

Moving the non-base models like `BertPretrainer` and `BertClassifier` to a new folder is fine, but won't they be exported to the same symbols anyway?

Reply:
How we self organize the code is a minor point. How we export the symbols is the big consideration here, and we fully control this.

My issue is our naming is far from self documenting. Here's our three symbols.
```
keras_nlp.models.Bert
keras_nlp.models.BertBase
keras_nlp.models.BertClassifier
```

No one would guess looking at these that `Bert` a generic configurable layer, `BertBase` is the only window to pretrained weights, and `BertClassifier` is an adapter for an instantiated model for a task.

The would be much more self documenting, but clunky.
```
keras_nlp.models.BertConfigurable
keras_nlp.models.BertBase
keras_nlp.models.BertClassificaitonAdapter
```

I am unsure what the best call is, so I won't hold up this PR any more. But this is an important thing to consider, we will quickly run our of runway to make major naming changes as people start using our stuff.

cc @fchollet in case you have thoughts on the naming here!

Reply:
IMO:

- Not sub-namespaces like `adapter`. They're all models.
- Agree that `Bert` vs `BertBase` is confusing, but that's just because the choice of "base" as a qualifier is confusing. It is expected by developers that BertBase is more abstract than Bert. But `BertLarge` vs `Bert` is not confusing. Simply picking a better name for `Bert`/`BertBase` (not sure -- should it be a descriptor of size?) could fix it.

Reply:
@fchollet @jbischof this?
```
keras_nlp.models.BertCustom
keras_nlp.models.BertBase
keras_nlp.models.BertLarge
keras_nlp.models.BertClassifier
```